autoload/lsp/ui/vim/semantic.vim

```diff
@@ -0,0 +1,194 @@
+let s:use_vim_textprops = has('textprop') && !has('nvim')
+let s:use_nvim_highlight = exists('*nvim_buf_add_highlight') && has('nvim')
+
+if s:use_nvim_highlight
+    let s:namespace_id = nvim_create_namespace('vim-lsp-semantic')
+endif
+
+if !hlexists('LspUnknownScope')
+    highlight LspUnknownScope gui=NONE cterm=NONE guifg=NONE ctermfg=NONE guibg=NONE ctermbg=NONE
+endif
+
+" Global functions {{{1
+function! lsp#ui#vim#semantic#is_enabled() abort
+    return g:lsp_semantic_enabled && (s:use_vim_textprops || s:use_nvim_highlight) ? v:true : v:false
+endfunction
+
+function! lsp#ui#vim#semantic#get_scopes(server) abort
+    if !lsp#capabilities#has_semantic_highlight(a:server)
+        return []
+    endif
+
+    let l:capabilities = lsp#get_server_capabilities(a:server)
+    return l:capabilities['semanticHighlighting']['scopes']
+endfunction
+
+function! lsp#ui#vim#semantic#handle_semantic(server, data) abort
+    if !g:lsp_semantic_enabled | return | endif
+
+    if lsp#client#is_error(a:data['response'])
+        call lsp#log('Skipping semantic highlight: response is invalid')
+        return
+    endif
+
+    let l:uri = a:data['response']['params']['textDocument']['uri']
+    let l:path = lsp#utils#uri_to_path(l:uri)
+    let l:bufnr = bufnr(l:path)
+
+    if l:bufnr < 0 | return | endif
```

Comment:
can we merge this check with `!bufloaded`? Or should `bufloaded` check be enough?

Reply:
Seems `bufloaded` already returns false if `bufnr` is `-1`, so only using `bufloaded` is enough.